### ISSUE TITLE
Graceful shutdown of kafka supervisor does not work properly

### ISSUE THREAD

**Issue by pjain1** [druid-io/druid#3495]
In case of graceful shutdown `workerExec` executor service is used to get the highest offset of tasks in the task group, set that as end offset for all tasks and resume them so that they finish and publish. However, since in most cases doing this will take more time than the time out for shutting down supervisor i.e. `SHUTDOWN_TIMEOUT_MILLIS`(15 seconds), `shutdownNow()` will be called on `workerExec`. Thus, when all the futures for pausing tasks resolves, further logic for getting highest offset and resuming them will not run and fail with exception like -
```
Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f with executor com.google.common.
util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f rejected from java.util.conc
urrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```
The paused tasks will do nothing and will never be finish.

**Comment by dclim** [comment]
@pjain1 have you seen this issue while shutting down a supervisor? I considered this possibility but I think that in most cases the shutdown logic should take less than 15 seconds unless the tasks are for some reason not very responsive (and there was an issue around this but #3401 should have fixed it)

**Comment by gianm** [comment]
Hmm, with graceful shutdowns, it's always good for there to be a timeout after which we give up and stop things forcefully. You never know when something will be stuck and not shutting down for some unforeseen reason, like a hardware problem, network problem, or a bug.

**Comment by dclim** [comment]
@pjain1 what if we made the supervisor shutdown timeout configurable + increased the default value, would that help alleviate the concern? That would allow users to set the timeout to something that makes sense for them while still retaining the hard kill to ensure that the supervisor doesn't leave any hung threads lying around

**Comment by pjain1** [comment]
Yes, I think that should work. FYI here's the log after graceful shutdown started - 
```
2016-09-21T21:42:18,133 INFO [qtp491762195-75] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Beginning shutdown of KafkaSupervisor[v2_metrics_cluster]
2016-09-21T21:42:18,133 INFO [qtp491762195-75] io.druid.indexing.overlord.RemoteTaskRunner - Unregistered listener [KafkaSupervisor-v2_metrics_cluster]
2016-09-21T21:42:18,133 INFO [qtp491762195-75] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Posting GracefulShutdownNotice, signalling managed tasks to complete and publish
2016-09-21T21:42:18,136 INFO [KafkaSupervisor-v2_metrics_cluster] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Task group [0] has run for [PT2100S]
2016-09-21T21:42:18,137 INFO [KafkaSupervisor-v2_metrics_cluster] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Task group [1] has run for [PT2100S]
2016-09-21T21:42:18,137 INFO [KafkaSupervisor-v2_metrics_cluster] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Task group [2] has run for [PT2100S]
2016-09-21T21:42:18,138 INFO [KafkaSupervisor-v2_metrics_cluster] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Task group [3] has run for [PT2100S]
2016-09-21T21:42:18,138 INFO [KafkaSupervisor-v2_metrics_cluster] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Task group [4] has run for [PT2100S]
2016-09-21T21:42:18,141 INFO [KafkaIndexTaskClient-v2_metrics_cluster-1] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm201.us.bf1.yahoo.com:8088
2016-09-21T21:42:18,144 INFO [KafkaIndexTaskClient-v2_metrics_cluster-3] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm200.us.bf1.yahoo.com:8088
2016-09-21T21:42:18,144 INFO [KafkaIndexTaskClient-v2_metrics_cluster-4] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-launcher-017.adx.bf1.yahoo.com:8088
2016-09-21T21:42:18,148 INFO [KafkaIndexTaskClient-v2_metrics_cluster-2] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm202.us.bf1.yahoo.com:8088
2016-09-21T21:42:18,152 INFO [KafkaIndexTaskClient-v2_metrics_cluster-0] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm203.us.bf1.yahoo.com:8089
2016-09-21T21:42:20,187 INFO [KafkaIndexTaskClient-v2_metrics_cluster-4] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-launcher-017.adx.bf1.yahoo.com:8088
2016-09-21T21:42:20,195 INFO [KafkaIndexTaskClient-v2_metrics_cluster-2] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm202.us.bf1.yahoo.com:8088
2016-09-21T21:42:20,198 INFO [KafkaIndexTaskClient-v2_metrics_cluster-4] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_0d3e79ec912ae94_cnidchdo] to pause;
 will try again in [PT2S]
2016-09-21T21:42:20,204 INFO [KafkaIndexTaskClient-v2_metrics_cluster-2] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_1d55fd6357389d4_hollblgo] to pause;
 will try again in [PT2S]
2016-09-21T21:42:20,467 INFO [KafkaIndexTaskClient-v2_metrics_cluster-1] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm201.us.bf1.yahoo.com:8088
2016-09-21T21:42:20,473 INFO [KafkaIndexTaskClient-v2_metrics_cluster-1] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_02610519a1559b7_ifmbcpgp] to pause;
 will try again in [PT2S]
2016-09-21T21:42:20,984 INFO [KafkaIndexTaskClient-v2_metrics_cluster-0] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm203.us.bf1.yahoo.com:8089
2016-09-21T21:42:20,998 INFO [KafkaIndexTaskClient-v2_metrics_cluster-0] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_be5537847f468f0_mifmfddl] to pause;
 will try again in [PT2S]
2016-09-21T21:42:22,204 INFO [KafkaIndexTaskClient-v2_metrics_cluster-4] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-launcher-017.adx.bf1.yahoo.com:8088
2016-09-21T21:42:22,210 INFO [KafkaIndexTaskClient-v2_metrics_cluster-2] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm202.us.bf1.yahoo.com:8088
2016-09-21T21:42:22,211 INFO [KafkaIndexTaskClient-v2_metrics_cluster-4] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_0d3e79ec912ae94_cnidchdo] to pause;
 will try again in [PT4S]
2016-09-21T21:42:22,218 INFO [KafkaIndexTaskClient-v2_metrics_cluster-2] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_1d55fd6357389d4_hollblgo] to pause;
 will try again in [PT4S]
2016-09-21T21:42:22,478 INFO [KafkaIndexTaskClient-v2_metrics_cluster-1] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm201.us.bf1.yahoo.com:8088
2016-09-21T21:42:22,485 INFO [KafkaIndexTaskClient-v2_metrics_cluster-1] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_02610519a1559b7_ifmbcpgp] to pause;
 will try again in [PT4S]
2016-09-21T21:42:23,004 INFO [KafkaIndexTaskClient-v2_metrics_cluster-0] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm203.us.bf1.yahoo.com:8089
2016-09-21T21:42:23,435 INFO [KafkaIndexTaskClient-v2_metrics_cluster-0] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_be5537847f468f0_mifmfddl] to pause;
 will try again in [PT4S]
2016-09-21T21:42:23,871 INFO [KafkaIndexTaskClient-v2_metrics_cluster-3] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm200.us.bf1.yahoo.com:8088
2016-09-21T21:42:23,887 INFO [KafkaIndexTaskClient-v2_metrics_cluster-3] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_f09a891f91d7499_dhfejhcc] to pause;
 will try again in [PT2S]
2016-09-21T21:42:25,894 INFO [KafkaIndexTaskClient-v2_metrics_cluster-3] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm200.us.bf1.yahoo.com:8088
2016-09-21T21:42:25,900 INFO [KafkaIndexTaskClient-v2_metrics_cluster-3] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_f09a891f91d7499_dhfejhcc] to pause;
 will try again in [PT4S]
2016-09-21T21:42:26,217 INFO [KafkaIndexTaskClient-v2_metrics_cluster-4] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-launcher-017.adx.bf1.yahoo.com:8088
2016-09-21T21:42:26,223 INFO [KafkaIndexTaskClient-v2_metrics_cluster-2] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm202.us.bf1.yahoo.com:8088
2016-09-21T21:42:26,224 INFO [KafkaIndexTaskClient-v2_metrics_cluster-4] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_0d3e79ec912ae94_cnidchdo] to pause;
 will try again in [PT8S]
2016-09-21T21:42:26,229 INFO [KafkaIndexTaskClient-v2_metrics_cluster-2] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_1d55fd6357389d4_hollblgo] to pause;
 will try again in [PT8S]
2016-09-21T21:42:26,490 INFO [KafkaIndexTaskClient-v2_metrics_cluster-1] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm201.us.bf1.yahoo.com:8088
2016-09-21T21:42:26,496 INFO [KafkaIndexTaskClient-v2_metrics_cluster-1] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_02610519a1559b7_ifmbcpgp] to pause; will try again in [PT8S]
2016-09-21T21:42:27,441 INFO [KafkaIndexTaskClient-v2_metrics_cluster-0] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm203.us.bf1.yahoo.com:8089
2016-09-21T21:42:27,492 INFO [KafkaIndexTaskClient-v2_metrics_cluster-0] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_be5537847f468f0_mifmfddl] to pause; will try again in [PT8S]
2016-09-21T21:42:29,906 INFO [KafkaIndexTaskClient-v2_metrics_cluster-3] com.metamx.http.client.pool.ChannelResourceFactory - Generating: http://druid-vm200.us.bf1.yahoo.com:8088
2016-09-21T21:42:29,913 INFO [KafkaIndexTaskClient-v2_metrics_cluster-3] io.druid.indexing.kafka.KafkaIndexTaskClient - Still waiting for task [index_kafka_v2_metrics_cluster_f09a891f91d7499_dhfejhcc] to pause; will try again in [PT8S]
2016-09-21T21:42:33,137 INFO [qtp491762195-75] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Timed out while waiting for shutdown
2016-09-21T21:42:33,137 INFO [qtp491762195-75] io.druid.indexing.kafka.supervisor.KafkaSupervisor - Shutdown notice handled
2016-09-21T21:42:33,137 INFO [qtp491762195-75] io.druid.indexing.kafka.supervisor.KafkaSupervisor - KafkaSupervisor[v2_metrics_cluster] has stopped
Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f with executor com.google.common.util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f rejected from java.util.concurrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@58e25739 with executor com.google.common.util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@58e25739 rejected from java.util.concurrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@36f471c with executor com.google.common.util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@36f471c rejected from java.util.concurrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@5d76bc6f with executor com.google.common.util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@5d76bc6f rejected from java.util.concurrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@33ff5c43 with executor com.google.common.util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@33ff5c43 rejected from java.util.concurrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

2016-09-21T21:42:33,142 ERROR [KafkaSupervisor-v2_metrics_cluster] io.druid.indexing.kafka.supervisor.KafkaSupervisor - KafkaSupervisor[v2_metrics_cluster] failed to handle notice: {class=io.druid.indexing.kafka.supervisor.KafkaSupervisor, exceptionType=class java.lang.InterruptedException, exceptionMessage=null, noticeClass=GracefulShutdownNotice}
java.lang.InterruptedException
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:998) ~[?:1.8.0_60]
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304) ~[?:1.8.0_60]
        at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:285) ~[guava-16.0.1.jar:?]
        at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116) ~[guava-16.0.1.jar:?]
        at io.druid.indexing.kafka.supervisor.KafkaSupervisor.checkTaskDuration(KafkaSupervisor.java:887) ~[druid-kafka-indexing-service-0.9.2-SNAPSHOT.jar:0.9.2-SNAPSHOT]
        at io.druid.indexing.kafka.supervisor.KafkaSupervisor.gracefulShutdownInternal(KafkaSupervisor.java:499) ~[druid-kafka-indexing-service-0.9.2-SNAPSHOT.jar:0.9.2-SNAPSHOT]
        at io.druid.indexing.kafka.supervisor.KafkaSupervisor$GracefulShutdownNotice.handle(KafkaSupervisor.java:463) ~[druid-kafka-indexing-service-0.9.2-SNAPSHOT.jar:0.9.2-SNAPSHOT]
        at io.druid.indexing.kafka.supervisor.KafkaSupervisor$2.run(KafkaSupervisor.java:299) [druid-kafka-indexing-service-0.9.2-SNAPSHOT.jar:0.9.2-SNAPSHOT]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_60]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_60]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]

```

**Comment by dclim** [comment]
@pjain1 okay cool I'll make the change

Interesting, any idea why the tasks were taking so long to pause (anything in the task logs)? Were they in the middle of a persist to disk?

**Comment by pjain1** [comment]
yes they were persisting and it took some time for them to finish the persist.  Will they not respond to commands from supervisor when they are persisting ?

**Comment by dclim** [comment]
@pjain1 hmm yeah, I believe if the persists are taking too long, and maxPendingPersists is exceeded, ingestion will be throttled, the run loop will block, and the pause command won't be handled immediately until the next run loop cycle.

**Comment by pjain1** [comment]
I see.. I think that's what I am running into. Thanks for the clarification.

**Comment by singhania** [comment]
Hi,

We are using druid 0.9.2 and once we submit shoutdown command to overlord for supervisor.It returns with supervisor id but doesn't stop the task.Task kept on running.

Also if we try to query about same supervisor we get supervisor doesn't exists.

I believe we are getting this issue but it should be fixed in 0.9.2.
